### DISCUSSION TITLE
Is there tooling to detect class collisions?

### DISCUSSION THREAD

**Discussion by bmulholland** [tailwindlabs/tailwindcss#10803] [Help]
There has [previously been discussion](https://github.com/tailwindlabs/tailwindcss/discussions/1446) on how to auto-merge conflicting classes (e.g. two backgrounds different background colors). One problem I see with that approach is: how do we tell which of the two is correct? While a decision can be made there, and that might be right, I propose that this implicit hierarchy is actually technical debt. For example, see CSS, where relying on specificity is often bad practice.

Today, having conflicting classes in our app has caused friction when updating Tailwind -- sometimes even in point versions. That's because the order of CSS output from Tailwind can change, which changes specificity, which changes the appearance of our app. I'd like to remove this source of friction and force us to be explicit about which classes we want to us.

There is linting that will detect conflicting classes statically, when used in the same component. The more complex cases are when the collisions happen due to a class outside a component conflicting with one inside of it, and specifically when those two are auto merged by e.g. vue.

I'd like a tool that detects these collisions and raises errors. It would probably have to be at runtime, such as a browser plugin. My ideal would be something automated that can be added to storybook and e2e tests, but anything is better than nothing. Does anything like that exist?

**Comment by louiss0** [comment]
> There has [previously been discussion](https://github.com/tailwindlabs/tailwindcss/discussions/1446) on how to auto-merge conflicting classes (e.g. two backgrounds different background colors). One problem I see with that approach is: how do we tell which of the two is correct? While a decision can be made there, and that might be right, I propose that this implicit hierarchy is actually technical debt. For example, see CSS, where relying on specificity is often bad practice.
> 
> Today, having conflicting classes in our app has caused friction when updating Tailwind -- sometimes even in point versions. That's because the order of CSS output from Tailwind can change, which changes specificity, which changes the appearance of our app. I'd like to remove this source of friction and force us to be explicit about which classes we want to us.
> 
> There is linting that will detect conflicting classes statically, when used in the same component. The more complex cases are when the collisions happen due to a class outside a component conflicting with one inside of it, and specifically when those two are auto merged by e.g. vue.
> 
> I'd like a tool that detects these collisions and raises errors. It would probably have to be at runtime, such as a browser plugin. My ideal would be something automated that can be added to storybook and e2e tests, but anything is better than nothing. Does anything like that exist?

You could try [this](https://www.npmjs.com/package/@code-fixer-23/cn-efs#guide)
